Source/inv.cpp

```diff
@@ -302,37 +302,23 @@ bool AutoEquip(Player &player, const Item &item, inv_body_loc bodyLocation, bool
 
 int FindSlotUnderCursor(Point cursorPosition, Size itemSize)
 {
-	int i = cursorPosition.x;
-	int j = cursorPosition.y;
-
+	Point hotSpot = cursorPosition;
 	if (!IsHardwareCursor()) {
 		// offset the cursor position to match the hot pixel we'd use for a hardware cursor
-		i += itemSize.width * INV_SLOT_HALF_SIZE_PX;
-		j += itemSize.height * INV_SLOT_HALF_SIZE_PX;
+		hotSpot += Displacement { itemSize.width * INV_SLOT_HALF_SIZE_PX, itemSize.height * INV_SLOT_HALF_SIZE_PX };
 	}
 
 	for (int r = 0; r < NUM_XY_SLOTS; r++) {
-		int xo = GetRightPanel().position.x;
-		int yo = GetRightPanel().position.y;
-		if (r >= SLOTXY_BELT_FIRST) {
-			xo = GetMainPanel().position.x;
-			yo = GetMainPanel().position.y;
-		}
+		Point panel = r < SLOTXY_BELT_FIRST ? GetRightPanel().position : GetMainPanel().position;
 
-		if (i >= InvRect[r].x + xo && i <= InvRect[r].x + xo + InventorySlotSizeInPixels.width) {
-			if (j >= InvRect[r].y + yo - InventorySlotSizeInPixels.height - 1 && j < InvRect[r].y + yo) {
-				return r;
-			}
-		}
-		if (r == SLOTXY_CHEST_LAST) {
-			if (itemSize.width % 2 == 0)
-				i -= INV_SLOT_HALF_SIZE_PX;
-			if (itemSize.height % 2 == 0)
-				j -= INV_SLOT_HALF_SIZE_PX;
-		}
-		if (r == SLOTXY_INV_LAST && itemSize.height % 2 == 0)
-			j += INV_SLOT_HALF_SIZE_PX;
+		Rectangle slot {
+			panel + Displacement { InvRect[r].x, InvRect[r].y - InventorySlotSizeInPixels.height },
+			InventorySlotSizeInPixels

```

Comment:
This also introduces an off-by one. The hit areas used in the old code are inclusive on both ends for the width. This is actually important because the definition of `InvRect` is pretty garbage, there's a line of pixels between the left and right columns of the weapon slots. With the old cold the hit area is enlarged by a single pixel closing this gap, this PR makes it possible to fail to equip a weapon by clicking on that line. The gutter between cells in the inventory and belt also becomes inactive unless you use `InventorySlotSizeInPixels + 1`.

The hit areas for body slots in general are a mess -_-, I ended up pulling out the proper regions and making #5925 for it, but this is what is currently covered by InvRect in master:
![2023-03-25T154617_devilutionx_markup](https://user-images.githubusercontent.com/357684/227701879-db6ba8d7-5ca9-465f-b0d6-2a0301eec0ed.png)

Apart from the gap in the middle of the weapon slots the ring areas are offset slightly left of where they should be, and head/body/right arm graphics are larger than the actual hit area.

Reply:
i did notice some of these issues yestaday when play testing some other tthings. in general there are still a lot that can be cleaned up about the inventory code :)